src/java.net.http/share/classes/jdk/internal/net/http/HttpConnection.java

```diff
@@ -76,17 +78,48 @@ abstract class HttpConnection implements Closeable {
     public static final Comparator<HttpConnection> COMPARE_BY_ID
             = Comparator.comparing(HttpConnection::id);
 
+    private static final AtomicLong LABEL_COUNTER = new AtomicLong();
+
     /** The address this connection is connected to. Could be a server or a proxy. */
     final InetSocketAddress address;
     private final HttpClientImpl client;
     private final TrailingOperations trailingOperations;
+
+    /**
+     * A unique identifier that provides a total order among instances.
+     */
     private final long id;
 
-    HttpConnection(InetSocketAddress address, HttpClientImpl client) {
+    /**
+     * A label to identify the connection.
+     * <p>
+     * This label helps with associating multiple components participating in a
+     * connection. For instance, an {@link AsyncSSLConnection} and the
+     * {@link PlainHttpConnection} it wraps will share the same label. As a
+     * result, compared to {@link #id}, this label does not give a total order
+     * among instances.
```

Comment:
Hello Volkan, I think we shouldn't talk about any kind of ordering here. I see that the `id` field has been updated with a comment stating that it provides ordering among instances. Even there, I don't think that comment about ordering is needed. So far we haven't used the `id` to mean anything other than a logical identifier and these values have played no role in ordering of connections at any place. So adding the comments about ordering, I think, makes it confusing.

Reply:
The only use of `HttpConnection.id` is for ordering - so that connections can be placed in a ConcurrentSkipListSet. So maybe here we should not speak of ordering (remove the last sentence in that paragraph) - but I think it helps to keep it for `id`.

Reply:
> The only use of HttpConnection.id is for ordering - so that connections can be placed in a ConcurrentSkipListSet.

I wasn't aware of that. Now that you mentioned it, I looked up the code which uses the Set to store these connections. And from what I can see, the order is only used during the closing of a HttpClient, to close these opened connections in that specific order. Did I miss any other usages of the order?

In any case, now that you corrected me about the usage of the `id` order, I agree that having the comment on the `id` field is fine and only remove it from the paragraph here.

Reply:
When we close a connection, we take it out of the set. So it's not about ordering the connection but about quickly finding the connection in that set.

Reply:
It was already a puzzler for me to figure out `id` cannot be used for `label`, and `label` cannot be used for `id` either. Both are due to the fact that `label` can be shared across instances, whereas `id` cannot. In short, I've found it very helpful to thoroughly document their these fields' properties and what they serve for. Unless you have a strong objection, I'd like to keep the current form of documentation.

Reply:
Hello Volkan, I think having the mention of ordering on the `id` field is OK. I still find it confusing to have this sentence on the `label`:

>
> As a result, compared to {@link #id}, this label does not give a total order among instances.

To me, this feels odd because `label` being a `String` typically wouldn't have been considered to be playing a role in ordering. Plus, in isolation, it's not clear what ordering this talks about (Daniel did explain to me that the `id` field is used in ordering of open connections. I haven't yet looked deeper at that code to see where that happens).

In any case, this is an internal class so if we do decide to use the current proposed text, that's OK with me.

Reply:
I agree with Jaikiran: I would just remove the last sentence in this pargraph.

Reply:
Implemented in 42ff0fc4a83cd33f1a35e4a2b3e7ce5c25ea39cc.